proposals/reward-proposal-v2-reward-calculation.md

```diff
@@ -0,0 +1,109 @@
+is proposed. Let's call this approach `V1`. A prototype of the `V1` proposal is
+implemented. While the `V1` approach works functionally, it, however, has two
+main drawbacks.
+
+First, a restriction is enforced to prevent updates to all stake accounts and
+vote accounts during reward interval. In other words, the stake accounts and
+vote accounts are **locked** during reward interval. All stake account's actions
+such as delegate, redelegate, withdraw etc. are disallowed during reward
+interval and have to wait after reward interval.
+
+Second, there is no double-entry booking available for epoch rewards. Currently,
+epoch rewards are rewarded and recorded inside the first block of the epoch.
+Lack of double-entry booking is less a problem for the current approach, because
+all rewards are contained one block - the first block of the epoch, which makes
+block reward verification somewhat easier and trackable. However, with
+distributed rewards (such as the `V1` approach ), lack of double-entry booking
+becomes a more serious limitation. Since the rewards distribution can span
+multiple blocks, it becomes more difficult to verify the correctness of the
+reward payment without the ability for double-entry booking.
+
+Given these two drawbacks above, we are proposing an improved way to handle
+epoch reward - `V2`. The `V2` approach is based on `V1` but with the
+improvements to address these two drawbacks.
+
+To make the discussion clearer, the `V2` approach is split into two proposals:
+`rewards calculation` and `rewards distribution`. In this proposal, we will
+focus on "reward calculation". A follow-up proposal will talk about "reward
+distribution".
+
+## Proposed Solutions
+
+Unlike the current epoch rewards computation approach, in which computing the
+rewards happens at the first slot of epoch boundary in the bank runtime
+directly, the new reward computation is moved to a background service in the
+proposed approach.
+
+This is very similar to the `V1` proposal, a separate service,
+"EpochRewardCalculationService" is created at the start time of the validator.
+This service is listening on a channel for incoming rewards calculation
+requests, and handle the requests to compute the rewards.
+
+There are two main differences between the `V2` and `V1` approaches.
+
+The first difference is that instead of sharing the `StakeCache` with the
+runtime during reward computation, the `StakeCache` is cloned at slot-height 0
+in the `V2` epoch. In this way, we can avoid locking stake accounts and vote
+accounts to prevent updates during the reward calculation.
+
+And the reward computation can last for the next `N` block-height. At the end of
+`Nth` block-height, the reward computation result should be available. In case
+that the reward result is not available yet at the `Nth` block-height, the
+runtime will wait until the computation is finished before processing any blocks
+with height greater than `N`.
```

Comment:
The reason for this fixed `N` is for consensus. Different nodes may run the computation at different speeds; thus, they can finish at different times. To make sure that every node pays the rewards at the same time, a convenient way to choose the fixed `N`. However, this doesn't have to be a fixed number. It can be a function computing the total number of rewards paying accounts.